docs/src/writing-constructs.rst

```diff
@@ -0,0 +1,208 @@
+.. Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+   This work is licensed under a Creative Commons Attribution-NonCommercial-ShareAlike 4.0
+   International License (the "License"). You may not use this file except in compliance with the
+   License. A copy of the License is located at http://creativecommons.org/licenses/by-nc-sa/4.0/.
+
+   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+   either express or implied. See the License for the specific language governing permissions and
+   limitations under the License.
+
+.. _writing_constructs:
+
+#########################
+Writing |cdk| Constructs
+#########################
+
+This topic provides some tips writing idiomatic new constructs for the |cdk|.
+The tips here apply equally to constructs written for inclusion in the AWS
+Construct Library, purpose-built constructs to achieve a well-defined goal,
+or constructs that serve as building blocks for assembling your cloud
+applications.
+
+General Design Priciples
+========================
+
+* Favor composition over inheritance; most of the constructs should directly
+  extend the **Construct** class instead of some other construct. Inheritance
+  should mainly be used to allow polymorphism. Typically, you'll add a child
+  construct and expose any of its APIs and properties in the parent construct.
+* Provide defaults for everything that a reasonable guess can be made for;
+  ideally, props should be optional and **new MyAwesomeConstruct(this, "Foo")**
+  should be enough to set up a reasonable variant of the construct. This does
+  not mean that the user should not have the opportunity to customize! Rather,
+  it means that the specific parameter should be optional and set to a
+  reasonable value if not supplied. This may involve creating other resources as
+  part of constructing this construct. For example, all resources that require a
+  role allow passing in a **Role** object (specifically, a **RoleRef** object),
+  but if the user does not supply one an appropriate **Role** object is
+  constructed in-place.
+* Use contextual defaulting between properties; the value of one property may
+  affect sensible defaults for other properties. *Example*:
+  **enableDnsHostnames**
+  and and **enableDnsSupport**. **dnsHostnames** requires **dnsSupport**, only
+  yell if the user has *explicitly* disabled DNS Support, but tried to enable
+  DNS Hostnames.  A user expects things to just work - make that the case unless
```

Comment:
The last part of the sentence ("make that the case") is a bit convoluted. Maybe something like "if the user intention was clear, there is no need to throw errors" or something like that